tests/catch/unit/memory/hipMemsetSync.cc

```diff
@@ -0,0 +1,505 @@
+      params.srcPtr = make_hipPitchedPtr(aPtr, dataPitch, dataW, dataH);
+      params.dstPtr = make_hipPitchedPtr(hostMem, dataW, dataW, dataH);
+
+      hipExtent extent;
+      extent.width = dataW;
+      extent.height = dataH;
+      extent.depth = dataD;
+
+      params.extent = extent;
+
+      HIP_CHECK(hipMemcpy3D(&params));
+      break;
+    }
+
+    case memSetType::hipMemset2D:
+      HIP_CHECK(hipMemcpy2D(hostMem, dataW * elementSize, aPtr, dataPitch, dataW, dataH,
+                            hipMemcpyHostToHost));
+      break;
+
+    default:
+      HIP_CHECK(hipMemcpy(hostMem, aPtr, sizeInBytes, hipMemcpyHostToHost));
+      break;
+  }
+}
+
+template <typename T>
+static void devRegisteredCopy(memSetType memType, T* aPtr, T* hostMem, size_t dataW, size_t dataH,
+                              size_t dataD, size_t& dataPitch) {
+  size_t elementSize = sizeof(T);
+
+  switch (memType) {
+    case memSetType::hipMemset3D: {
+      hipMemcpy3DParms params{};
+      params.kind = hipMemcpyHostToHost;
+      params.srcPos = make_hipPos(0, 0, 0);
+      params.dstPos = make_hipPos(0, 0, 0);
+      params.srcPtr = make_hipPitchedPtr(aPtr, dataPitch, dataW, dataH);
+      params.dstPtr = make_hipPitchedPtr(hostMem, dataW, dataW, dataH);
+
+      hipExtent extent;
+      extent.width = dataW;
+      extent.height = dataH;
+      extent.depth = dataD;
+
+      params.extent = extent;
+
+      HIP_CHECK(hipMemcpy3D(&params));
+      break;
+    }
+
+    case memSetType::hipMemset2D:
+      HIP_CHECK(hipMemcpy2D(hostMem, dataW * elementSize, aPtr, dataPitch, dataW, dataH,
+                            hipMemcpyDeviceToHost));
+      break;
+
+    default: {
+      size_t sizeInBytes = elementSize * dataW * dataH * dataD;
+      HIP_CHECK(hipMemcpy(hostMem, aPtr, sizeInBytes, hipMemcpyDeviceToHost));
+      break;
+    }
+  }
+}
+
+// Copies device data to host and checks that each element is equal to the
+// specified value
+template <typename T>
+void verifyData(T* aPtr, size_t value, MultiDData& data, allocType type, memSetType memType) {
+  auto dataH = data.height == 0 ? 1 : data.height;
+  auto dataD = data.depth == 0 ? 1 : data.depth;
+  std::unique_ptr<T[]> hostPtr = std::make_unique<T[]>(data.pitch * dataH * dataD / sizeof(T));
+  switch (type) {
+    case allocType::deviceMalloc:
+      deviceMallocCopy(memType, aPtr, hostPtr.get(), data.width, dataH, dataD, data.pitch);
+      break;
+    case allocType::devRegistered:
+      devRegisteredCopy(memType, aPtr, hostPtr.get(), data.width, dataH, dataD, data.pitch);
+      break;
+    default:  // host allocated or host registered memory
+      hostCopy(memType, aPtr, hostPtr.get(), data.width, dataH, dataD, data.pitch);
+      break;
+  }
+
+
+  size_t idx;
+  bool allMatch = true;
+
+  for (size_t k = 0; k < dataD; k++) {
+    for (size_t j = 0; j < dataH; j++) {
+      for (size_t i = 0; i < data.width; i++) {
+        idx = data.pitch * dataH * k + data.pitch * j + i;
```

Comment:
again I think we should only be looking at the extent here so I think it would be:
```suggestion
        idx = data.width * dataH * k + data.width * j + i;
```
please double check yourself